File: agentmail-gtm-agent/python/prospects.py
"""
CSV-backed prospect tracker.

prospects.csv has the following columns (header row required):
  email, name, role, company, hook, status, first_touch_at,
  followup_at, replied_at, classification, thread_id

Drop new prospects in by editing the CSV. The agent picks up rows where
`status` is empty or 'queued' on its next polling cycle.
"""

import csv
from datetime import datetime, timezone
from pathlib import Path

PROSPECTS_FILE = Path("prospects.csv")
LOG_FILE = Path("gtm_log.csv")

COLUMNS = [
    "email", "name", "role", "company", "hook",
    "status", "first_touch_at", "followup_at",
    "replied_at", "classification", "thread_id",
]

LOG_COLUMNS = [
    "timestamp_utc", "action", "prospect_email", "classification",
    "thread_id", "note",
]


# --- prospects --------------------------------------------------------------


def load_all() -> list[dict]:
    if not PROSPECTS_FILE.exists():
        return []
    with PROSPECTS_FILE.open(newline="", encoding="utf-8") as f:
        return [_normalize(r) for r in csv.DictReader(f)]


def _normalize(row: dict) -> dict:
    """Ensure all expected columns exist, defaulting to empty string."""
    return {col: (row.get(col) or "").strip() for col in COLUMNS}


def save_all(rows: list[dict]) -> None:
    with PROSPECTS_FILE.open("w", newline="", encoding="utf-8") as f:
        w = csv.DictWriter(f, fieldnames=COLUMNS)
        w.writeheader()
        for r in rows:
            w.writerow({c: r.get(c, "") for c in COLUMNS})


def update_prospect(email: str, **fields) -> dict | None:
    rows = load_all()
    target = None
    for r in rows:
        if r["email"].lower() == email.lower():
            r.update(fields)
            target = r
            break
    if target:
        save_all(rows)
    return target


def find_by_thread(thread_id: str) -> dict | None:
    for r in load_all():
        if r["thread_id"] == thread_id:
            return r
    return None


def queued_prospects() -> list[dict]:
    return [r for r in load_all() if r["status"] in ("", "queued")]


def followups_due(after_hours: int) -> list[dict]:
    """Prospects who got a first touch >N hours ago and haven't been followed up
    or replied. Returns rows where status == 'first_touch_sent'."""
    cutoff = datetime.now(timezone.utc).timestamp() - after_hours * 3600
    out = []
    for r in load_all():
        if r["status"] != "first_touch_sent":
            continue
        if not r["first_touch_at"]:
            continue
        try:
            ts = datetime.fromisoformat(r["first_touch_at"]).timestamp()
        except Exception:
            continue
        if ts <= cutoff:
            out.append(r)
    return out


# --- log --------------------------------------------------------------------


def _ensure_log_header() -> None:
    if not LOG_FILE.exists() or LOG_FILE.stat().st_size == 0:
        with LOG_FILE.open("w", newline="", encoding="utf-8") as f:
            csv.writer(f).writerow(LOG_COLUMNS)


def log_action(*, action: str, prospect_email: str, classification: str = "",
               thread_id: str = "", note: str = "") -> None:
    _ensure_log_header()
    row = [
        datetime.now(timezone.utc).isoformat(timespec="seconds"),
        action,
        prospect_email,
        classification,
        thread_id,
        note.replace("\n", " ")[:500],
    ]
    with LOG_FILE.open("a", newline="", encoding="utf-8") as f:
        csv.writer(f).writerow(row)

File: agentmail-gtm-agent/typescript/src/prospects.ts
/**
 * CSV-backed prospect tracker + audit log.
 */

import * as fs from "node:fs";

const PROSPECTS_FILE = "prospects.csv";
const LOG_FILE = "gtm_log.csv";

const COLUMNS = [
  "email", "name", "role", "company", "hook",
  "status", "first_touch_at", "followup_at",
  "replied_at", "classification", "thread_id",
] as const;

const LOG_COLUMNS = [
  "timestamp_utc", "action", "prospect_email", "classification",
  "thread_id", "note",
] as const;

export interface Prospect {
  email: string;
  name: string;
  role: string;
  company: string;
  hook: string;
  status: string;
  first_touch_at: string;
  followup_at: string;
  replied_at: string;
  classification: string;
  thread_id: string;
}

// --- CSV parsing helpers (RFC 4180 minimal) ---------------------------------

function csvEscape(s: string): string {
  if (s == null) return "";
  if (/[",\n]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
  return s;
}

function parseCsvLine(line: string): string[] {
  const out: string[] = [];
  let cur = "";
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (inQuotes) {
      if (c === '"' && line[i + 1] === '"') { cur += '"'; i++; }
      else if (c === '"') inQuotes = false;
      else cur += c;
    } else {
      if (c === '"') inQuotes = true;
      else if (c === ",") { out.push(cur); cur = ""; }
      else cur += c;
    }
  }
  out.push(cur);
  return out;
}

function readCsv(path: string): Record<string, string>[] {
  if (!fs.existsSync(path)) return [];
  const text = fs.readFileSync(path, "utf8");
  const lines = text.split(/\r?\n/).filter((l) => l.length);
  if (!lines.length) return [];
  const headers = parseCsvLine(lines[0]);
  const rows: Record<string, string>[] = [];
  for (let i = 1; i < lines.length; i++) {
    const cells = parseCsvLine(lines[i]);
    const row: Record<string, string> = {};
    headers.forEach((h, idx) => { row[h] = (cells[idx] ?? "").trim(); });
    rows.push(row);
  }
  return rows;
}

function writeCsv(path: string, rows: Record<string, string>[], columns: readonly string[]): void {
  const header = columns.join(",");
  const body = rows.map((r) => columns.map((c) => csvEscape(r[c] ?? "")).join(",")).join("\n");
  fs.writeFileSync(path, header + "\n" + (body ? body + "\n" : ""));
}

function normalize(row: Record<string, string>): Prospect {
  const out: any = {};
  for (const c of COLUMNS) out[c] = (row[c] ?? "").trim();
  return out as Prospect;
}

// --- prospects --------------------------------------------------------------

export function loadAll(): Prospect[] {
  return readCsv(PROSPECTS_FILE).map(normalize);
}

export function saveAll(rows: Prospect[]): void {
  writeCsv(PROSPECTS_FILE, rows as any, COLUMNS);
}

export function updateProspect(email: string, fields: Partial<Prospect>): Prospect | undefined {
  const rows = loadAll();
  let target: Prospect | undefined;
  for (const r of rows) {
    if (r.email.toLowerCase() === email.toLowerCase()) {
      Object.assign(r, fields);
      target = r;
      break;
    }
  }
  if (target) saveAll(rows);
  return target;
}

export function findByThread(threadId: string): Prospect | undefined {
  return loadAll().find((r) => r.thread_id === threadId);
}

export function queuedProspects(): Prospect[] {
  return loadAll().filter((r) => r.status === "" || r.status === "queued");
}

export function followupsDue(afterHours: number): Prospect[] {
  const cutoff = Date.now() - afterHours * 3600 * 1000;
  return loadAll().filter((r) => {
    if (r.status !== "first_touch_sent") return false;
    if (!r.first_touch_at) return false;
    const ts = new Date(r.first_touch_at).getTime();
    return !isNaN(ts) && ts <= cutoff;
  });
}

// --- log --------------------------------------------------------------------

function ensureLogHeader(): void {
  if (!fs.existsSync(LOG_FILE) || fs.statSync(LOG_FILE).size === 0) {
    fs.writeFileSync(LOG_FILE, LOG_COLUMNS.join(",") + "\n");
  }
}

export function logAction(opts: {
  action: string;
  prospect_email: string;
  classification?: string;
  thread_id?: string;
  note?: string;
}): void {
  ensureLogHeader();
  const row = [
    new Date().toISOString(),
    opts.action,
    opts.prospect_email,
    opts.classification || "",
    opts.thread_id || "",
    (opts.note || "").replace(/\n/g, " ").slice(0, 500),
  ].map(csvEscape).join(",");
  fs.appendFileSync(LOG_FILE, row + "\n");
}
